design/general_refactorings.md

```diff
@@ -0,0 +1,381 @@
+| host           | files related to host command     |
+| init           | files related to init command     |
+| list           | files related to list command     |
+| login          | files related to login command    |
+| logout         | files related to logout command   |
+| policy         | files related to policy command   |
+| user           | files related to user command     |
+| variable       | files related to variable command |
+
+The following files are without parent directory : api.py, client.py, argparse_wrapper.py, cli.py , client.py, config.py, constants.py, credentials_data.py, credentials_form_file.py, endpoint.py, errors.py, http_wrapper.py, resource.py, ssl_service.py, version.py
+
+##### Recommendation
+
+I recommend that we organize the files by their types and not by the commands they represent.
+
+| Directory Name | Purpose                                                      |
+| :------------- | :----------------------------------------------------------- |
+| wrappers       | wrappers and utils of diffrent python libs :  arg_parse_wrapper, http_wrapper |
+| conjur_api     | files that talk directly with conjur. For Example   `conjur_client.py (conjur.py) , conjur_api.py (api.py), ssl_client.py (ssl_service)` |
+| logic          | all logic classes                                            |
+| controller     | all controller classes                                       |
+| logic          | all logic classes                                            |
+| data_objects   | all data objects classes                                     |
+
+
+
+### Improve Error Handling
+
+------
+
+##### Proposal
+
+This will help to better under the error handling flow. This will allow us to catch our own object and not general exceptions that do not proper helpful information to the end user. All Conjur CLI/SDK-specific exceptions should be documented in the funtions that call them.
+
+##### Current state
+
+Through the codebase, we have general errors that are not representative of the actual error that took place. The exception is not documented in the fucntion documentation in the code.
+For example, in the following `get_certificate` function, we raise a general exception if anything goes wrong during the certificate retrieval operation.
+
+```python
+def get_certificate(self, hostname, port):
+    """
+    Method for connecting to Conjur to fetch the certificate chain
+    """
+    if port is None:
+        port = DEFAULT_PORT
+    try:
+        fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)
+        logging.debug("Successfully fetched certificate")
+    except Exception as error:
+        raise Exception(f"Unable to retrieve certificate from {hostname}:{port}. " \
+                        f"Reason: {str(error)}") from error
+
+    return fingerprint, readable_certificate
+```
+
+We raise Exception if anything happens in certificate retrivial.
```

Comment:
why is that bad? Please add here why

Reply:
Its not . I think i copy pasted from appendix . Removing this line